### PR TITLE
Add claim REST endpoint

### PR DESCRIPTION
**Pull request by neilbmclaughlin**
This is a basic implementation of just the `/payment-schedule` endpoint to return all payment schedules with no error handling.

### CODE REVIEW

File: server/services/schedule-service.js [hunk 1]
```diff
@@ -2,7 +2,22 @@ const scheduleRepository = require('../repository/schedule-repository')
 
 const BASE_PAYMENTS = 6
 
+function aggregateByClaimId (scheduleLines) {
```
Comment (neilbmclaughlin) [inline server/services/schedule-service.js]: This method is a candidate to be refactored for simplicity but it works.

File: server/services/schedule-service.js [hunk 2]
```diff
@@ -2,7 +2,22 @@ const scheduleRepository = require('../repository/schedule-repository')
...
+
+  return Object
+    .entries(schedules)
+    .map((e) => { return { claimId: e[0], paymentDates: e[1] } })
+}
+
 module.exports = {
+  getAll: async function () {
+    const scheduleLines = await scheduleRepository.getAll()
+    return aggregateByClaimId(scheduleLines)
```
Comment (markysoft) [inline server/services/schedule-service.js]: Would this be better done in SQL rather than a getAll passed to node, i.e.
```
SELECT "claimId", array_to_string(array_agg("paymentDate"), ',')
FROM schedules
GROUP BY "claimId";
```

File: test/unit/server/services/schedule-service-retrieve.test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      }
+    ]
+
+    scheduleRepository.getAll = jest.fn(() => repositoryResponse)
+
+    const retrieveSchedules = await scheduleService.getAll()
+
+    expect(retrieveSchedules).toEqual(expectedSchedules)
+  })
+})
```
Comment (markysoft) [inline test/unit/server/services/schedule-service-retrieve.test.js]: I get the impression this test is largely verifying that the function `aggregateByClaimId` is transforming the `getAll` results into the `expectedSchedules` structure, so could be covered by a test on that function alone and reduce the amount of mocking. 

Also if the aggregate logic is performed in SQL, then this test could be removed altogether.

File: package.json [hunk 4]
```diff
@@ -10,7 +10,7 @@
     "start:watch": "nodemon --inspect=0.0.0.0 --ext js --legacy-watch index.js",
     "test": "npm run test:unit",
     "test:lint": "standard",
-    "test:unit": "jest --forceExit",
+    "test:unit": "jest --forceExit --detectOpenHandles",
```
Comment (markysoft) [inline package.json]: did you mean to keep the `detectOpenHandles` in? I think this it to assist in diagnosing if there are functions still running

File: server/services/payment-service.js [hunk 5]
```diff
@@ -1,5 +1,13 @@
 const paymentRepository = require('../repository/payment-repository')
 
+function paymentMapper (payment) {
+  return {
+    claimId: payment.claimId,
+    paymentAmount: Number.parseFloat(payment.value).toFixed(2),
```
Comment (markysoft) [inline server/services/payment-service.js]: There's a lot of weird edge cases where rounding doesn't work as you'd expect in javascript. Not suggesting any changes here, but for a real payment system we'd need tests to ensure all the rounding work as we expect.

File: server/services/payment-service.js [hunk 6]
```diff
@@ -9,5 +17,15 @@ module.exports = {
     }
     console.log('creating payment')
     await paymentRepository.create(calculation)
+  },
+  getAll: async function () {
+    const payments = await paymentRepository.getAll()
+    const returnValue = payments.map(paymentMapper)
+    return returnValue
```
Comment (markysoft) [inline server/services/payment-service.js]: You could return the result of `payments.map` rather than assigning it to a variable, then returning it. This would be consistent with the approach taken in the `schedule-service`

File: server/services/payment-service.js [hunk 7]
```diff
@@ -9,5 +17,15 @@ module.exports = {
...
+  },
+  getById: async function (claimId) {
+    const payment = await paymentRepository.getById(claimId)
+    const returnValue = paymentMapper(payment)
+    return returnValue
```
Comment (markysoft) [inline server/services/payment-service.js]: As above, the `returnValue` variable could be removed.

File: test/integration/routes/payment.test.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    await db.payment.bulkCreate([
+      { claimId: 'MINE123', value: 150.50 },
+      { claimId: 'MINE124', value: 50.75 }
+    ])
+  })
+
+  beforeEach(async () => {
+    server = await createServer()
+    await server.initialize()
+    jest.clearAllMocks()
```
Comment (markysoft) [inline test/integration/routes/payment.test.js]: Looks like there aren't any mocks in the test so this line could be removed.

File: test/integration/routes/payment.test.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    await db.schedule.bulkCreate([
+      { scheduleId: 1, claimId: 'MINE123', paymentDate: '2020-03-01 14:30' },
+      { scheduleId: 2, claimId: 'MINE123', paymentDate: '2020-04-01 14:30' },
+      { scheduleId: 3, claimId: 'MINE124', paymentDate: '2020-05-01 14:30' }
+    ])
+    await db.payment.bulkCreate([
+      { claimId: 'MINE123', value: 150.50 },
+      { claimId: 'MINE124', value: 50.75 }
+    ])
+  })
```
Comment (markysoft) [inline test/integration/routes/payment.test.js]: I think it might make the test cleaner if some of the sample data is held in variables, i.e.
`const claim123 = { claimId: 'MINE123', value: 150.50 }`
So the expectations are against an existing constant, rather than a newly defined constant.

It's a matter of personal preference really, and there's an added complication of the date format being slightly different for the `bulkCreate`

File: test/integration/routes/schedule.test.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    await db.payment.bulkCreate([
+      { claimId: 'MINE123', value: 150.50 },
+      { claimId: 'MINE124', value: 50.75 }
+    ])
+  })
+
+  beforeEach(async () => {
+    server = await createServer()
+    await server.initialize()
+    jest.clearAllMocks()
```
Comment (markysoft) [inline test/integration/routes/schedule.test.js]: This line could be removed as there are no mocks.

File: server/services/payment-service.js (hunk 6 above)
Comment (neilbmclaughlin) [reply server/services/payment-service.js]: This must be a common error pattern resulting from the addition and removal of `console.log(...)` statements.

File: package.json (hunk 4 above)
Comment (neilbmclaughlin) [reply package.json]: I added this at the suggestion of jest and to reduce the warning noise when doing a test run.

File: test/integration/routes/payment.test.js (hunk 9 above)
Comment (neilbmclaughlin) [reply test/integration/routes/payment.test.js]: I'm going to leave this as is for now - I prefer my inputs and expected outputs to be separate as it feels more like double entry book-keeping.

Maybe we can use it as a COP discussion point.

File: server/services/schedule-service.js (hunk 2 above)
Comment (neilbmclaughlin) [reply server/services/schedule-service.js]: The change to using the ORM means the aggregation is no longer needed.

File: test/unit/server/services/schedule-service-retrieve.test.js (hunk 3 above)
Comment (neilbmclaughlin) [reply test/unit/server/services/schedule-service-retrieve.test.js]: The change to using the ORM means the aggregation is no longer needed.
